### PR TITLE
Initial support for export-hook

### PR DESCRIPTION
**Pull request by milessabin**
Once this is in Kittens will build and we'll have a clearer picture of how this is playing out.

### CODE REVIEW

**Comment by milessabin** [comment]
All snapshot dependencies now replaced with releases, so this is ready to merge now.

File: core/src/main/scala/alleycats/ConsK.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+object ConsK extends ConsK0 {
+  implicit def pureSemigroupKIsConsK[F[_]](implicit p: Pure[F], s: SemigroupK[F]): ConsK[F] =
+    new ConsK[F] {
+      def cons[A](hd: A, tl: F[A]): F[A] = s.combine(p.pure(hd), tl)
+    }
+}
+
+@imports[ConsK]
+trait ConsK0
```
Comment (ceedubs) [inline core/src/main/scala/alleycats/ConsK.scala:19]: Should this be `private[alleycats] sealed trait ConsK0`? I don't know if that causes any issues with export hook. But in general I think that if a trait just exists as an implementation detail then we shouldn't expose it for extension/usage by users of the library. If nothing else, making it private makes it obvious to users that they aren't meant to try to use it.

Comment (ceedubs) [inline core/src/main/scala/alleycats/ConsK.scala:19]: This comment applies to various other places in this PR as well.

File: build.sbt [hunk 2]
```diff
@@ -16,7 +16,14 @@ addCommandAlias("gitSnapshots", ";set version in ThisBuild := git.gitDescribedVe
 val gh = GitHubSettings(org = "non", proj = "alleycats", publishOrg = "org.typelevel", license = mit)
 val devs = Seq(Dev("Erik Osheim", "non"))
 
-val vers = typelevel.versions ++ alleycats.versions
+val updates = Map(
+  "macro-compat" -> "1.1.0",
+  "export-hook" -> "1.1.0",
+  "simulacrum" -> "0.5.0"
+)
+val updatesSettings = Seq( resolvers += Resolver.sonatypeRepo("snapshots"))
```
Comment (ghost) [inline build.sbt]: > All snapshot dependencies now replaced with releases, so this is ready to merge now.

Small point; As you now have no shapshot dependendencies, might we an idea to remove/comment out this line?

File: build.sbt [hunk 3]
```diff
@@ -86,7 +93,7 @@ lazy val commonSettings = sharedCommonSettings ++
   addCompilerPlugins(vAll, "kind-projector") ++ Seq(
   scalacOptions ++= scalacAllOptions,
   parallelExecution in Test := false
-) ++ warnUnusedImport ++ unidocCommonSettings
+) /* ++ warnUnusedImport */ ++ unidocCommonSettings // spurious warnings from macro annotations expected
```
Comment (ghost) [inline build.sbt:95]: Small point (again): locally this works for me enabled, perhaps worth re-adding?

File: core/src/main/scala/alleycats/ConsK.scala (hunk 1 above)
Comment (milessabin) [inline core/src/main/scala/alleycats/ConsK.scala:19]: My default position is that all pure code should be public unless there's a compelling reason to the contrary. I don't think there is here, so I wouldn't make this private as a matter of course.

File: build.sbt (hunk 2 above)
Comment (milessabin) [inline build.sbt]: Agreed.

File: build.sbt (hunk 3 above)
Comment (milessabin) [inline build.sbt:95]: Oh! Interesting ... I'll try that again.

Comment (milessabin) [inline build.sbt:95]: I'm still seeing the unused import warnings :-(

Comment (ghost) [inline build.sbt:95]: Sorry for this, so am I  :-(

File: core/src/main/scala/alleycats/ConsK.scala (hunk 1 above)
Comment (ceedubs) [inline core/src/main/scala/alleycats/ConsK.scala:19]: We have a huge scala infrastructure at work, so fighting binary incompatibilities in library versions is usually my compelling reason to not expose something that is kind of an internal detail :)

Comment (milessabin) [inline core/src/main/scala/alleycats/ConsK.scala:19]: Agreed that there's a discussion to be had here which goes beyond the scope of this PR. Specifically though, I don't see any binary incompatibility issues arising from this trait.

**Comment by non** [comment]
Side-stepping the larger issues around public/private, this PR seems good to me. I'd be happy to merge it and then change stuff later if/when we decide it is necessary. :+1:

**Comment by ceedubs** [comment]
:+1:

**Pull request merged**